src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1461 @@
+        self.fc2 = nn.Linear(config.encoder_ffn_dim, self.embed_dim)
+        self.final_layer_norm = LayerNorm(self.embed_dim)
+
+    def forward(self, x, encoder_padding_mask, output_attentions=False):
+        residual = x
+        x, attn_weights = self.self_attn(
+            query=x, key=x, key_padding_mask=encoder_padding_mask, output_attentions=output_attentions
+        )
+        x = F.dropout(x, p=self.dropout, training=self.training)
+        x = residual + x
+        x = self.self_attn_layer_norm(x)
+        residual = x
+        x = self.activation_fn(self.fc1(x))
+        x = F.dropout(x, p=self.activation_dropout, training=self.training)
+        x = self.fc2(x)
+        x = F.dropout(x, p=self.dropout, training=self.training)
+        x = residual + x
+        x = self.final_layer_norm(x)
+        return x, attn_weights
+
+
+
+class ProphetNetEncoder(nn.Module):
+    """
+    ProphetNet Encoder is same to traditional Transformer Encoder.
+    Transformer encoder consisting of *config.encoder_layers* self attention layers. Each layer
+    is a :class:`EncoderLayer`.
+
+    Args:
+        config: ProphetNetConfig
+    """
+    def __init__(self, config: ProphetNetConfig, embed_tokens):
+        super().__init__()
+
+        self.output_hidden_states = config.output_hidden_states
+
+        self.dropout = config.dropout
+        embed_dim = embed_tokens.embedding_dim
+        self.padding_idx = embed_tokens.padding_idx
+        self.max_source_positions = config.max_position_embeddings
+
+        self.embed_tokens = embed_tokens
+        self.embed_scale = None
+        self.embed_positions = LearnedPositionalEmbedding(config.max_position_embeddings+1+self.padding_idx, embed_dim, self.padding_idx)
+
+        self.layers = nn.ModuleList([])
+        self.layers.extend([
+            EncoderLayer(config)
+            for i in range(config.encoder_layers)
+        ])
+        self.emb_layer_norm = LayerNorm(embed_dim)
+
+    def forward(self, input_ids, attention_mask=None, output_attentions=False):
+        """
+
+        Args:
+            input_ids (LongTensor): tokens in the source language of shape
+                `(batch, src_len)`
+            attention_mask (torch.LongTensor): indicating which indices are padding tokens.
+        Returns:
+            Tuple comprised of:
+                - **x** (Tensor): the last encoder layer's output of
+                  shape `(src_len, batch, embed_dim)`
+                - **encoder_states** (List[Tensor]): all intermediate
+                  hidden states of shape `(src_len, batch, embed_dim)`.
+                  Only populated if *output_hidden_states:* is True.
+                - **all_attentions** (List[Tensor]): Attention weights for each layer.
+                During training might not be of length n_layers because of layer dropout.
+
+        """
+
+        # remove bos to be consistent with fairseq version
+        input_ids = input_ids[:, 1:]
```

Comment:
This seems strange -> does that mean the first token of the `input_ids` is useless? Can we maybe change that even though it will be different to fairseq

Reply:
This is what I suggested the author to do. The pretrained weights are without the first token (bos) so there would be no other way to validate those pretrained weights. As long as all inputs are without bos I think it's no big deal. @patrickvonplaten

Reply:
I don't fully follow here sorry. It would be nice if we don't have to restrain the model to *not* have BOS token as the first token.

Reply:
@patrickvonplaten  I think you may be missing something here. So let me elaborate a little. This is because the authors used `fairseq` to train the models originally. In `fairseq`, you can set a flag to strip the BOS of the input:
```
input: Patrick is happy. [/s]
output: [s] Patrick is happy. [/s]
```
But in our library, we always have a `[s]` in both the input and output. This creates a mismatch between the training and this would compromise the performance of uploaded weights. Please refer to the **edit history** of the PR description (I think @qiweizhen edit that out after modifying as I suggested). 

In a nutshell, the model drops the encoded `[s]` at the beginning. I still think it's totally okay to do so.

Reply:
Or we can think of something from the tokenizer side. However @qiweizhen and I tried but found it's not feasible